### PR TITLE
BUG: DataFrame[int] +/- datetime64

### LINKED ISSUE #28080: BUG: Series[int] + datetime64 should raise
identified [here](https://github.com/pandas-dev/pandas/commit/a6c0d02c99cfaf90119d1316dd897019326472c1#r34783276).  We handle this correctly for timedelta64, but apparently not datetime64

```
iser = pd.Series(range(3))
dt = np.datetime64(3, "D")

>>> iser + dt
0   1970-01-04
1   1970-01-05
2   1970-01-06
```

Should behave like `iser + pd.Timestamp(dt)`:

```
>>> iser + pd.Timestamp(dt)
[...]
ValueError: Cannot add integral value to Timestamp without freq.
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #28080 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/ops/__init__.py [hunk 1]
```diff
@@ -149,13 +149,24 @@ def maybe_upcast_for_op(obj, shape: Tuple[int, ...]):
...
     """
-    from pandas.core.arrays import TimedeltaArray
+    from pandas.core.arrays import DatetimeArray, TimedeltaArray
 
     if type(obj) is datetime.timedelta:
         # GH#22390  cast up to Timedelta to rely on Timedelta
         # implementation; otherwise operation against numeric-dtype
         # raises TypeError
         return Timedelta(obj)
+    elif isinstance(obj, np.datetime64):
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: can this ever be called with a ``datetime.datetime`` (or subclass eg Timestamp),?

**Comment by jreback** [comment]
this is technically a bug fix? can you add a whatsnew note

File: pandas/core/ops/__init__.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: datetime and Timestamp both behave correctly already, also are included in the test

**Comment by jbrockmendel** [comment]
whatsnew note added, and green

File: pandas/tests/arithmetic/test_numeric.py [hunk 2]
```diff
@@ -235,6 +235,30 @@ def test_add_sub_timedeltalike_invalid(self, numeric_idx, other, box):
...
+    @pytest.mark.parametrize(
+        "other",
+        [
+            pd.Timestamp.now().to_pydatetime(),
+            pd.Timestamp.now(tz="UTC").to_pydatetime(),
+            pd.Timestamp.now().to_datetime64(),
+            pd.NaT,
+        ],
+    )
+    @pytest.mark.filterwarnings("ignore:elementwise comp:DeprecationWarning")
```
Comment (TomAugspurger) [inline pandas/tests/arithmetic/test_numeric.py]: Who / what is this from?

Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_numeric.py]: in ensure_python_int we have a 'assert new_value == value` (inside a try/except) that numpy is warning about when `value` is a np.datetime64 obj

Comment (TomAugspurger) [reply pandas/tests/arithmetic/test_numeric.py]: This raises for me with NumPy 1.17

```python

In [13]: a = np.datetime64(10, 'ns')

In [14]: pd.core.dtypes.common.ensure_python_int(a)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
~/sandbox/pandas/pandas/core/dtypes/common.py in ensure_python_int(value)
    203         new_value = int(value)
--> 204         assert new_value == value
    205     except (TypeError, ValueError, AssertionError):

AssertionError:

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-14-a45e0cb68b99> in <module>
----> 1 pd.core.dtypes.common.ensure_python_int(a)

~/sandbox/pandas/pandas/core/dtypes/common.py in ensure_python_int(value)
    204         assert new_value == value
    205     except (TypeError, ValueError, AssertionError):
--> 206         raise TypeError(msg.format(type(value), value))
    207     return new_value
    208

TypeError: Wrong type <class 'numpy.datetime64'> for value 1970-01-01T00:00:00.000000010
```

It just warned earlier? If so, then +1

Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_numeric.py]: > This raises for me with NumPy 1.17

Yah, the assertion raises on master, shouldn't be affected by this PR.  This PR makes `pd.Series([0]) + pd.Timestamp.now().to_datetime64()` raise TypeError, which doesn't happen now.

**Comment by jbrockmendel** [comment]
@jreback gentle ping; this is a blocker for the ops-perf fixing

**Review by TomAugspurger — APPROVED** [review]

**Pull request merged**